### ISSUE TITLE
[Reading] Training nGPT

### ISSUE THREAD

**Issue by ShotaSato01** [ShotaSato01/Paper_survey#144]
## 0. Paper
https://arxiv.org/abs/2608.01284

## 1. What is it?
正規化Transformer (nGPT)は、モデルのパラメータベクトルと活性化ベクトルを単位超球面上に制約することで、「超球面表現学習」を実現する。
本論文では、nGPTを実際に学習するための実用的なレシピを提示し、最新のハイブリッドmamba-e-Transformer Mixture-of-Expertモデル上で評価する

この学習レシピでは、ロジット勾配事前調整、対数的学習率減衰、GatedAdamW、角度更新の制御、および任意で利用可能な探索機構を導入する

同一のハイブリッドMoEアーキテクチャを用い、AdamWで学習した非正規化モデルと比較したところ、総パラメータ数140億のnGPTモデルは、およそ半分の学習トークン数で同じ検証損失に到達した。

さらに、この学習レシピは、本論文で検討した最大総パラメータ数140億のモデルまで、一貫してスケールすることが確認された。

<img width="1270" height="789" alt="Image" src="https://github.com/user-attachments/assets/231d56bc-88b5-4b9c-a1a9-4b3837135263" />
  
1) Attentionブロックで、青い点として表された二つのトークン「Life」「is」を考慮し、緑色の点で示される予測候補を提示する。
2) この予測に対して、隠れ状態と組み合わさって、当たらな中間点が生成される。
3) 次に、この中間点がMLPブロックに渡され、MLPは赤い点で示される別の予測候補を提示する。
4) 隠れ状態は、MLPが提示した方向へ2回目の学習済みステップを踏む

→このプロセスを層ごとに繰り返すことで、次トークン予測は、超球面上での最適化プロセスとして捉えることができる。
モデルのベクトルはアンカーとして機能し、ないせきはそれらのアンカーとの類似度を測定し、AttentionブロックとMLPブロックが順次、次に進むべきステップを提案する。

本研究で検討するnGPT学習レシピは、以下の要素から構成される
・Logit Gradient Preconditioning：ロジット勾配事前調整
・Logarithmic Learning Rate Decay：対数的学習率減衰
・GatedAdamW
・Pre-Moment Tangent Projection：モーメント計算前の接空間射影
・Angular Step Cap：角度ステップ上限
・Second-Moment Growth Clipping：2次モーメント増大クリッピング
・Post-Moment Exploration Noise：モーメント計算後の探索ノイズ

## 3. Where is the key to technologies and techniques?

<img width="1506" height="989" alt="Image" src="https://github.com/user-attachments/assets/956b1b19-046f-4ee6-946c-74efaef9ecba" />

<img width="1550" height="697" alt="Image" src="https://github.com/user-attachments/assets/70729a6d-5266-4506-866a-228c97fa8c46" />

<img width="1067" height="831" alt="Image" src="https://github.com/user-attachments/assets/df03df87-fa4b-49ba-9ed1-cfc6d7edd487" />
  
→εが「数値の安定化」と「小さい2次モーメントの更新抑制」という二つの仕事を同時にこなさないようにする。

### Section 3: Training nGPT

Section 3 では、hybrid Mamba-2–Transformer MoE を **実際に nGPT として学習するための具体的な変換レシピ**をまとめている。:contentReference[oaicite:0]{index=0}

中心的な考え方は、**パラメータや hidden state の方向は unit hypersphere 上に制約し、必要な大きさは trainable scale として別に持たせる**こと。

主要な parameter vector は

$$
\|w\|_2 = 1
$$

となるよう正規化し、residual update も

$$
h \leftarrow \mathrm{Norm}\left(h+\alpha\odot(h_{\text{block}}-h)\right)
$$

のように毎回 unit norm に戻す。一方で、正規化だけでは Attention や GELU/SwiGLU、Mamba、MoE router に必要なスケールまで失われるため、

$$
x_{\text{effective}} = s\odot \mathrm{Norm}(x)
$$

という形で、$q,k$、MLP/MoE の中間表現、Mamba 入力、MoE router 入力、logit などに学習可能な scaling factor を追加する。:contentReference[oaicite:1]{index=1} :contentReference[oaicite:2]{index=2}

実際の nGPT 化では、RMSNorm / LayerNorm と weight decay を削除し、GatedAdam、Pre-Moment Tangent Projection、Angular Step Cap、LGP、PMEN、SMGC などを組み合わせる。また、optimizer update 後には対象パラメータを再度 unit norm に正規化する。:contentReference[oaicite:3]{index=3}

要するに、

>**主要な parameter vector と activation を unit norm に正規化し、Attention / MLP の更新量は学習可能な eigen learning rate で制御する。一方、正規化によって失われるスケールは、$q,k$、MLP/MoE 中間表現、Mamba 入力、MoE router、logit に個別の trainable scale を導入して補う。
さらに RMSNorm / LayerNorm と weight decay を除去し、GatedAdam、Pre-Moment Tangent Projection、Angular Step Cap、LGP、PMEN、SMGC を適用して、optimizer update 後に対象パラメータを再び unit norm に正規化する。**

#### 4.1 Experimental Setup

実験では、Nemotron Team が開発した **Nemotron-3 Nano 系の hybrid Mamba–Transformer MoE** を使用し、1B / 2B / 4B / 7B / 14B の5サイズで scaling を評価する。

各モデルの総パラメータ数と active parameter / token、学習トークン数は以下。

- 1B: 0.21B active, 88B tokens
- 2B: 0.37B active, 140B tokens
- 4B: 0.61B active, 215B tokens
- 7B: 0.91B active, 310B tokens
- 14B: 1.74B active, 560B tokens

baseline の非正規化モデルを **GPT**、正規化版を **nGPT** と呼ぶ。

GPT baseline は AdamW を用い、

$$
\beta_1 = 0.9,\quad \beta_2 = 0.95,\quad \lambda = 0.1
$$

とする。1B tokens の warmup 後、モデルサイズに応じて peak learning rate を下げ、WSD schedule で peak の 1% まで減衰させる。

一方 nGPT では、提案した logarithmic learning rate decay を

$$
\rho = 0.05
$$

で使用し、peak learning rate は

$$
\frac{0.24}{\sqrt{d_{\mathrm{model}}}}
$$

とする。

また、expert parameter の learning rate は 1.5 倍、router は 2 倍にする。$s_z$ と一部の Mamba parameter は最初の 10% を warmup し、Angular Step Cap も同期間で

$$
0^\circ \rightarrow 1^\circ
$$

まで増加させる。

GatedAdam の設定は

$$
\epsilon_{\mathrm{gate}} = 10^{-8},\quad
\epsilon_{\mathrm{num}} = 10^{-14},\quad
\beta_1 = \beta_2 = 0.975
$$

とする。

Second-Moment Growth Clipping は 500 step 後から

$$
R = 100
$$

で有効化する。ただし、$R=2$ や clipping 無効でも最終性能はほぼ同じであり、SMGC は必須要素というより **稀な gradient spike に対する safety mechanism** と位置づけられている。

#### 4.2 Experimental Results

**Figure 5 / Table 1** では、1B〜14B の各モデルサイズについて GPT + AdamW と nGPT + GatedAdam の training / validation loss を比較している。nGPT は全サイズで一貫して低い loss を示し、validation loss は概ね **約 2.5%改善**している。

<img width="979" height="653" alt="Image" src="https://github.com/user-attachments/assets/5f059b59-c4fe-4be1-a7ec-497c6bd2d19e" />

<img width="1012" height="377" alt="Image" src="https://github.com/user-attachments/assets/e1ef094c-7d3f-4a8c-bc9a-e1f7584f0266" />

**Figure 6** では、4B / 7B / 14B モデルについて、GatedAdam が利用する second-moment scale

$$
\sqrt{\hat v}
$$

の分布と gate の関係を示している。モデルが大きくなるほど $`sqrt{\hat{v}}`$ の中央値は小さくなり、小さい $`sqrt{\hat{v}}`$ を持つ座標では gate が強く効く。$`a=0.5`$ とすることで、$`sqrt{\hat{v}} < \epsilon_{\mathrm{gate}}`$ の座標でも $a=1$ より大きな更新を与えられる。

<img width="904" height="732" alt="Image" src="https://github.com/user-attachments/assets/b89c3e2c-3af4-4ce5-9259-57feb9a8a37d" />

**Figure 7** では、14B モデルの loss-token curve を比較している。nGPT は序盤の loss 低下は GPT より遅いが、後半ではより速く収束し、GPT + AdamW と同じ training / validation loss に **約半分の training tokens** で到達する。

<img width="907" height="532" alt="Image" src="https://github.com/user-attachments/assets/40a12c56-3ea9-406b-ae79-183dba2c29dd" />

**Figure 8** では、nGPT 内で GatedAdam の

$$
a=0.5
$$

と Adam 相当の

$$
a=1
$$

を比較している。$a=0.5$ の方が良い結果を示し、同じ validation loss に到達するまでに **約 15〜20%少ない training tokens** で済む。ただし、この改善は GPT → nGPT 全体の差よりかなり小さい。

<img width="949" height="596" alt="Image" src="https://github.com/user-attachments/assets/94a3f2bf-044e-4e9a-8203-5567dd84cc89" />

したがって、Figure 5–8 の結果を合わせると、nGPT の性能向上は **GatedAdam 単体ではなく、parameter normalization、scaling、learning-rate schedule、optimizer modification を含む training recipe 全体**によるものと考えられる。

Pre-Moment Tangent Projection と PMEN の ablation はそれぞれ **Appendix B / Figure 9**、**Appendix D / Figure 10** で評価されており、validation loss への効果は比較的小さい。

## 7. personal comment 
LGPによって何が嬉しいか？
・「設定した学習率」と「実際に聞いている学習率」がずれにくくなる
・勾配ノルムの不要な増大を抑えやすい
・「出力をどれだけ尖らせたいか」と「ネットワーク全体をどれくらい大きく更新したいか」を、同じパラメータに勝手に兼任させなくてよくなる

**Fig.6**

> 大きなモデルでは $\sqrt{\hat{v}}$ が小さい座標が多い  
> ↓  
> $a=1$ だと gate がほぼ閉じる座標が多い  
> ↓  
> $a=0.5$ にして、そういう座標も少し更新しよう

**Fig.8**

> 実際に $a=0.5$ にしたらどうなった？  
> ↓  
> $a=1$ より少し良くなった  
> ↓  
> ただし nGPT 全体の改善を説明するほど大きな効果ではない
